### PR TITLE
TLS/SSL custom log - v2

### PR DESCRIPTION
**Pull request by fooinha**
Custom format output for TLS/SSL

Includes a Common custom logging format.

Rebased.
Replaces #2449

### CODE REVIEW

File: doc/userguide/output/custom-tls-logging.rst [hunk 1]
```diff
@@ -0,0 +1,46 @@
+Custom tls logging
+===================
+
+As of Suricata 3.2.0 you can enable a custom tls logging option.
```
Comment (inliniac) [inline doc/userguide/output/custom-tls-logging.rst]: No need to mention a suricata version here. The docs are per version.

File: src/log-tlslog.c [hunk 2]
```diff
@@ -273,14 +320,119 @@ static OutputCtx *LogTlsLogInitCtx(ConfNode *conf)
...
+                }
+                break;
+            case LOG_TLS_CF_EXTENDED:
+            /* Extended format  */
+                LogTlsLogExtended(aft, ssl_state);
+                break;
+            default:
+            /* NO MATCH */
+                MemBufferWriteString(aft->buffer, LOG_CF_NONE);
+                SCLogDebug("No matching parameter %%%c for custom http log.", node->type);
```
Comment (inliniac) [inline src/log-tlslog.c]: "http log" should be "tls log"

File: src/log-cf-common.c [hunk 3]
```diff
@@ -0,0 +1,283 @@
...
+#include "util-unittest.h"
+
+/**
+ *  \brief Creates a custom format node
+ *  \retval LogCustomFormatNode * ptr if created
+ *  \retval NULL if failed to allocate
+ */
+LogCustomFormatNode * LogCustomFormatNodeAlloc()
+{
+    LogCustomFormatNode * node = SCMalloc(sizeof(LogCustomFormatNode));
```
Comment (inliniac) [inline src/log-cf-common.c]: SCCalloc? Then the memset on node below can go away.

File: src/log-cf-common.c [hunk 4]
```diff
@@ -0,0 +1,283 @@
...
+    if (unlikely(node == NULL)) {
+        SCLogError(SC_ERR_MEM_ALLOC, "Failed to alloc custom format node");
+        return NULL;
+    }
+    memset(node, '\0', sizeof(LogCustomFormatNode));
+    memset(node->data, '\0', LOG_NODE_STRLEN);
```
Comment (inliniac) [inline src/log-cf-common.c]: Seems redundant?

File: src/log-cf-common.c [hunk 5]
```diff
@@ -0,0 +1,283 @@
...
+}
+
+/**
+ *  \brief Creates a custom format.
+ *  \retval LogCustomFormat * ptr if created
+ *  \retval NULL if failed to allocate
+ */
+LogCustomFormat * LogCustomFormatAlloc()
+{
+    LogCustomFormat * cf = SCMalloc(sizeof(LogCustomFormat));
```
Comment (inliniac) [inline src/log-cf-common.c]: SCCalloc?

File: src/log-cf-common.c [hunk 6]
```diff
@@ -0,0 +1,283 @@
...
+
+
+#ifdef UNITTESTS
+/**
+ * \internal
+ * \brief This test tests default timestamp format
+ */
+static int LogCustomFormatTest01(void)
+{
+
```
Comment (inliniac) [inline src/log-cf-common.c]: style: plz remove this newline

File: src/util-time.c [hunk 7]
```diff
@@ -208,6 +208,15 @@ void CreateUtcIsoTimeString (const struct timeval *ts, char *str, size_t size)
     }
 }
 
+void CreateFormattedTimeString (const struct tm *t, const char *fmt, char *str, size_t size)
+{
+    if (likely(t != NULL) && likely(fmt != NULL) && likely(str != NULL) ) {
```
Comment (inliniac) [inline src/util-time.c]: I think you can simply wrap the whole condition in likely() instead of per subcondition

**Comment by fooinha** [comment]
Replace by #2643 - Review changes.

Thanx @inliniac

**Pull request closed without merging**